Add node tuning and monitoring guide (#3322)
OffchainLabs/arbitrum-docs: docs/run-arbitrum-node/nitro/node-tuning-and-monitoring.mdx

### Formula

```
GOMEMLIMIT = Container_Memory_Limit - Non_Go_Memory - Safety_Margin
```

You should use a safety margin of 300-500 MB to account for allocator overhead, transient allocations, and kernel page cache.

### Example: 16 GB container with defaults

| Component              | Size         | Source                             |
| ---------------------- | ------------ | ---------------------------------- |
| Pebble block cache     | 2,048 MB     | `database-cache` (CGO)             |
| Pebble memtables (max) | 1,024 MB     | `database-cache / 2` (CGO)         |
| Trie-clean cache       | 600 MB       | `trie-clean-cache` (fastcache)     |
| Snapshot cache         | 400 MB       | `snapshot-cache` (fastcache)       |
| Stylus WASM cache      | 256 MB       | `stylus-lru-cache-capacity` (Rust) |
| Malloc arenas          | 128 MB       | `MALLOC_ARENA_MAX=2`               |
| Thread stacks          | 300 MB\*     | ~2 MB per thread                   |
| **Total non-Go**       | **4,756 MB** |                                    |

> \*Thread stack usage depends on the number of active threads, which varies by workload.

Reviewer: Needs validation

For the main Nitro node, the Helm chart does not override `GOMAXPROCS`, letting Go's runtime auto-detect from the container CPU limit.

## Metrics and monitoring

Nitro exposes Prometheus-compatible metrics through a dedicated metrics server.

### Enable the metrics server

Pass the `--metrics` flag when starting the node. Configure the server with these flags:

| Flag                               | Default     | Description                              |
| ---------------------------------- | ----------- | ---------------------------------------- |
| `--metrics-server.addr`            | `127.0.0.1` | Listen address for the metrics server    |
| `--metrics-server.port`            | `6070`      | Listen port for the metrics server       |
| `--metrics-server.update-interval` | `3s`        | How often internal metrics are refreshed |

Example:

```shell
nitro --metrics \
  --metrics-server.addr 0.0.0.0 \
  --metrics-server.port 6070
```

### Prometheus scrape endpoint

Once metrics are enabled, Nitro exposes a Prometheus-compatible endpoint at:

```
http://<metrics-server.addr>:<metrics-server.port>/debug/metrics/prometheus
```

> The same server also exposes pprof profiling endpoints at `/debug/pprof/` for CPU and memory profiling.

Reviewer 2: ```suggestion
`pprof` runs as a separate server at `/debug/pprof/` for CPU and memory profiling, enabled with `--pprof` and configured via `--pprof-cfg.addr / --pprof-cfg.port` (default `127.0.0.1:6071`). It is not served on the metrics port.
```

## Tuning recommendations

### General rules

1. **Set `MALLOC_ARENA_MAX=2`**: This is the single most impactful change for containerized nodes. Without it, glibc can waste gigabytes on arena overhead, causing RSS to drift upward over days. Set this environment variable on every Nitro container.

2. **Start from the formula**: Calculate `GOMEMLIMIT` using the [formula above](#formula) with your actual cache configuration values. Do not set it to the container memory limit.

3. **Monitor RSS, not just Go heap**: Set container memory alerts based on actual RSS (`container_memory_rss` in Prometheus / cAdvisor), not Go-reported memory.

4. **All caches are bounded**: Unlike memory leaks, all non-Go memory in Nitro is bounded by configuration. With `MALLOC_ARENA_MAX` set, if RSS is stable and predictable, the node is behaving correctly. The memory is simply allocated outside Go's visibility.

### Example configurations

#### Full node with 32 GB RAM

Use default cache sizes. Calculate `GOMEMLIMIT` from the non-Go budget:

```shell
# Non-Go: 2048 + 1024 + 600 + 400 + 256 + 128 + 300 = 4,756 MB
# GOMEMLIMIT = 32,768 - 4,756 - 500 safety ≈ 27,512 MB ≈ 27 GB

> MALLOC_ARENA_MAX=2

Reviewer 2: ```suggestion
MALLOC_ARENA_MAX=2 \
```

> GOMEMLIMIT=27512MiB

Reviewer 2: ```suggestion
GOMEMLIMIT=27512MiB \
```
